pkg/defaults/defaults.go

```diff
@@ -397,9 +406,44 @@ func leasesForTest(s *api.MultiStageTestConfigurationLiteral) (ret []api.StepLea
 	return
 }
 
+// rootImageResolver creates a resolver for the root image import step. We attempt to resolve the root image and
+// the build cache. If we are able to successfully determine that the build cache is up-to-date, we import it as
+// the root image.
+func rootImageResolver(client loggingclient.LoggingClient, ctx context.Context) func(root, cache *api.ImageStreamTagReference) (*api.ImageStreamTagReference, error) {
+	return func(root, cache *api.ImageStreamTagReference) (*api.ImageStreamTagReference, error) {
+		cacheTag := &imagev1.ImageStreamTag{}
+		if err := client.Get(ctx, ctrlruntimeclient.ObjectKey{Namespace: cache.Namespace, Name: fmt.Sprintf("%s:%s", cache.Name, cache.Tag)}, cacheTag); err != nil {
+			if kapierrors.IsNotFound(err) {
+				// no build cache, use the normal root
+				return root, nil
+			}
+			return nil, fmt.Errorf("could not resolve build cache image stream tag %s: %w", cache.ISTagName(), err)
+		}
+		metadata := &docker10.DockerImage{}
+		if len(cacheTag.Image.DockerImageMetadata.Raw) == 0 {
+			return nil, fmt.Errorf("could not fetch Docker image metadata build cache %s", cache.ISTagName())
```

Comment:
Same here -> Is there a way how we easily discover problems like this without failing jobs for the user whose job would be fine if we didn't try to be smart?

Reply:
I thought about making this entirely best-effort but yes as you note it means that failures will be silent and hard to detect.

I also don't know that this really increases the burden on the job for success. We have three things that can fail here:

1. two GET ISTag calls
2. a JSON unmarshal

One of the two GET calls are calls we'd be doing anyway when we resolve `root`, so we only add the cost of potentially flaking here. I don't know that we really see flakes in the API server for our clusters, they are either up or not.

The JSON unmarhsal can never fail in a working system. I don't think it's possible for the build API to produce bad metadata here in the first place.

IMHO as-is gives us the best ratio of usefulness/discoverability/etc.

Reply:
Makes sense!